Added vignette highlight to gate showcases

MrJackey opened the pull request:
## Summary
Fixes 190
Whenever a gate is opened/closed they are highlighted to increase visibility.

Commit:
Added vignette highlight to gate showcases

smeas commented on Assets/VignetteHighlight.cs:
```diff
@@ -0,0 +1,61 @@
+		Canvas canvas = GetComponent<Canvas>();
+
+		if (canvas == null) {
+			Debug.LogError($"{nameof(VignetteHighlight)} requires a canvas");
```
You can put a `[RequireComponent(typeof(Canvas))]` attribute on the class to make sure that you can't add the component to a game object without a canvas. (Only checks when adding a component from the editor)

smeas commented on Assets/VignetteHighlight.cs:
```diff
@@ -0,0 +1,61 @@
+
+	private void FadeHighlight() {
+		alpha = MathX.EaseInQuad(0, baseAlpha, fadeTime / fadeDuration);
+		Mathf.Clamp(alpha, 0, baseAlpha);
```
You're not using the return value of the function.
Also, do you even need to clamp it? You're making sure that `fadeTime` wont go past `fadeDuration` on line 38.

smeas commented on Assets/VignetteHighlight.cs:
```diff
@@ -0,0 +1,61 @@
+		alpha = MathX.EaseInQuad(0, baseAlpha, fadeTime / fadeDuration);
+		Mathf.Clamp(alpha, 0, baseAlpha);
+
+		material.color = new Color(material.color.r, material.color.g, material.color.b, alpha);
```
You're getting the `color` property of the material three times here. How about instead of storing just the `baseAlpha`, you store the whole `baseColor`?

smeas commented on Assets/Scripts/Puzzle/Gate.cs:
```diff
@@ -52,14 +53,17 @@ public class Gate : SignalReceiver {
 
-	private IEnumerator CoShowEvent() {
+	private IEnumerator CoShowEvent(VignetteHighlight highlight) {
+		highlight.WorldTarget = transform;
```
Would be nicer to do this just below the instantiation. Maybe move it in here too?

MrJackey replied:
Would be nice yes but since they are using the same highlight I left it here to skip having to keep track and check if the highlight is instantiated already.

smeas reviewed (changes requested):
EaseOut generally looks better for a fade in and EaseIn for a fade out. Maybe try that? You might have to add a slight delay before starting the fading for it to look good though.

Addehman reviewed (commented):
Was it in this branch that the gate by the switch in the tutorial level, the one that doesn't have any way to be opened, wasn't this supposed to be fixed here too? 
Else, it looks fantastic! think the vignette is at a perfect alpha level too!

Commit:
Made suggested changes

Moved files to better folder.
Changed fading from easein & easeout -> easeout & easein

smeas reviewed (approved):
Awesome!

Addehman reviewed (approved):
Beautiful!
